### ISSUE TITLE
ENABLE_TESTABILITY requirement for dependency 

### ISSUE THREAD

**Issue by cameroncooke** [CocoaPods/CocoaPods#8437]
# Report

I have an issue where I'm unable to validate and therefore deploy a private podspec due to a requirement that a dependency is compiled with `ENABLED_TESTABILITY=YES` flag.

## Example podspec

```ruby
Pod::Spec.new do |s|
  s.name         = "MyTestingPod"
  s.version      = "1.3"
  s.summary      = "Bla"
  s.description  = "Bla bla bla"
  s.ios.deployment_target = "11.0"
  s.swift_version = "4.2"
  s.source        = { :git => "git@github.com:company-here/MyTestingPod.git", :tag => s.version.to_s }
  s.source_files  = "Sources/**/*"
  s.dependency 'SomeDependency', "1.3"
end
```

I've simplified the podspec but essentially `MyTestingPod` has a dependency of `SomeDependency`.

`MyTestingPod` is to be used by iOS Unit Test targets and sources in `MyTestingPod` use `@testable import SomeDependency`.

The issue is that as CocoaPods generates the Xcode project for the purposes of validation of the Podspec, there is no obvious way to require the `SomeDependency` is compiled with `ENABLED_TESTABILITY=YES` so that the files compile correctly.

It seems completely reasonable that a pod could be used by only test targets and that sources import a dependency using `@testable`. 

## What did you do?

`pod repo push --allow-warnings MySpecs MyTestingPod.podspec`

## What did you expect to happen?

I would expect the pod to validate and push to my private spec repo.

## What happened instead?

```
    - ERROR | [iOS] xcodebuild:  MyTestingPod/Sources/SomeFile.wift:7:18: error: module 'SomeDependency' was not compiled for testing.
```

### Stack

```
   CocoaPods : 1.5.3
        Ruby : ruby 2.3.4p301 (2017-03-30 revision 58214) [x86_64-darwin17]
    RubyGems : 2.5.2
        Host : Mac OS X 10.14.2 (18C54)
       Xcode : 10.1 (10B61)
         Git : git version 2.18.0
Ruby lib dir : /Users/cameron/.rbenv/versions/2.3.4/lib
Repositories : master - https://github.com/CocoaPods/Specs.git @ 9de7b8e0b8d36c6b1ff9097b96c8c41cd2c01f7d
               company_here - git@github.com:company_here/PodSpecs.git @ 9740d88e453122efcb653026f0c609ca10570ac0
               OurPrivateSpecs - git@github.com:company_here/PodSpecs.git @ 9740d88e453122efcb653026f0c609ca10570ac0
```

### Installation Source

```
Executable Path: /Users/cameron/.rbenv/versions/2.3.4/bin/pod
```

### Plugins

```
cocoapods-deintegrate : 1.0.2
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.0.0
cocoapods-trunk       : 1.3.1
cocoapods-try         : 1.1.0
```

**Comment by dnkoutso** [comment]
Sorry this does not appear to be a CocoaPods library issue. Please use StackOverflow under `cocoapods` tag for such issues.

**Comment by cameroncooke** [comment]
Why so quick to close?

Of course it’s a library issue the library doesn’t support what I’m trying to do and my point is that it clearly should be possible.

**Comment by cameroncooke** [comment]
If you look at my issue I state that I’m unable to deploy my private pod due to the requirement that the library validates the pod first. 

There is no way to specify that a dependency built when this library generates the project for validation that it is built with ‘ENABLE_TESTABILITY=YES’. 

This is causing my pod to incorrectly fail validation which I believe therefore is a bug or a shortfall in the library design.

**Comment by dnkoutso** [comment]
@cameroncooke I searched the codebase and I cant' find a place in which CocoaPods sets this value. What is your podspec?

**Comment by cameroncooke** [comment]
Let me try to explain again.

Say I've got a pod called `SausageTestHelpers` which is designed to provide support classes for unit tests to test code against another pod let's call it `Sausage`.

The purpose of the `SausageTestHelpers` cocoapod is to expose factories and helpers for testing the classes defined in `Sausage`.

The setup is simple, `SausageTestHelpers` has a dependency on `Sausage`. 

Example pod specs.

Sausage.podspec
```ruby
Pod::Spec.new do |s|
  s.name = "Sausage"
end
```

SausageTestHelpers.podspec
```ruby
Pod::Spec.new do |s|
  s.name = "SausageTestHelpers"
  s.dependency 'Sausage'
end
```

Let's imagine `Sausage` has some internal methods and let imagine `SausageTestHelpers` needs access to them, well, in that case, you would ensure that when `SausageTestHelpers` imports the `Sausage` module you prefix the import statements with @testable, for example:

```swift
@testable import Sausage

class MyTestHelper {
  func doSomethingWithInternalMethodsOfSausage() {
    let sausage = Sausage()
    sausage.doSomethingWithThisInternalMethod()
  }
}
```

The issue as I've pointed out, is that I can't get the `SausageTestHelpers` pod to pass validation due to a compiler error. The reason for the compiler error is that Cocoapods *doesn't* add the `ENABLED_TESTABILITY=YES` build setting to the project that is auto-generated. This is required for swift sources to compile when using `@testable`.

My requirement in this instance is that the auto-generated project that is created to run the podspec validation include this build setting. Without it, the validation will fail due to build errors.

When using these two pods from another project using the development pods and manually setting `ENABLED_TESTABILITY=YES` everything works as expected. But there is no way to control how the validation project is configured.

This means that I'm unable to bypass validation when pushing the perfectly valid Pod and podspec to our private repository.

While it's not a code-level issue I've added examples to try and make my user case clearer.

**Comment by dnkoutso** [comment]
Have you looked into `pod_target_xcconfig` which allows you to set build settings?

```ruby
spec.pod_target_xcconfig = { 'ENABLED_TESTABILITY' => 'YES' }
```

Sorry if I closed the issue too quickly, lets continue the discussion here and I will re-open if we can find this is a bug or enhancement.

**Comment by dnkoutso** [comment]
And if you mean this is on the sample app then you are right this is not added to it. You can still use `spec.user_target_xcconfig` instead but its not generally recommended.

**Comment by dnkoutso** [comment]
From Xcodeproj it seems that the default value that gets set to the project is set to `YES` https://github.com/CocoaPods/Xcodeproj/blob/master/lib/xcodeproj/constants.rb#L386

**Comment by cameroncooke** [comment]
Ok let me look into the further and see what’s going on.

**Comment by cameroncooke** [comment]
So here's the thing.

I manually commented out the validation lines in `push.rb` just to get my private pod to push to our private spec repo. Everything works now and the pods work as expected.

I've since restored the `push.rb` back to its original state and tried pushing again to test whether the configuration options you suggested would fix the issue moving forward as I don't want to be editing Cocoapods source every time and now can't reproduce the error.

Not sure what's changed or what has fixed it but it's no longer complaining about the build not being compiled for testing.

Thanks for your help, and it's good to know that we can add build settings in the podspecs as this was something I was unsure about.

**Comment by cameroncooke** [comment]
Just an update in-case anyone else encounters this issue, adding `spec.pod_target_xcconfig = { 'ENABLE_TESTABILITY' => 'YES' }` to the pod that should be complied for testing fixes this issue as it adds the build setting `ENABLE_TESTABILITY` to all build configurations.

**Comment by Oni-zerone** [comment]
Hi, I have the same issue, even adding the `spec.pod_target_xcconfig = { 'ENABLE_TESTABILITY' => 'YES' }`...
This is my podspec, maybe someone could help me here:
```ruby
Pod::Spec.new do |s|
    s.name             = 'PowerToolsTester'
    s.version          = '0.4.1'
    s.summary          = 'Tools tester for fast and modular app development.'
    s.swift_version    = '5.0'
    
    # This description is used to generate tags and improve search results.
    #   * Think: What does it do? Why did you write it? What is the focus?
    #   * Try to keep it short, snappy and to the point.
    #   * Write the description between the DESC delimiters below.
    #   * Finally, don't worry about the indent, CocoaPods strips it!
    
    s.description      = <<-DESC
    Fast development tools with a better way to use UICollectionViews
    DESC
    
    s.homepage         = 'https://github.com/Oni-zerone/PowerTools'
    s.license          = { :type => 'MIT', :file => 'LICENSE' }
    s.author           = { 'Andrea Altea' => 'oni.zerone@gmail.com' }
    s.source           = { :git => 'https://github.com/Oni-zerone/PowerTools.git', :tag => s.version.to_s + '-tester' }
    s.social_media_url = 'https://twitter.com/Oni_zerone'
    s.user_target_xcconfig = { 'ENABLED_TESTABILITY' => 'YES' }
    
    s.ios.deployment_target = '9.0'
    
    s.default_subspec = 'CoreTester'
    s.subspec 'CoreTester' do |sp|
        sp.source_files = 'PowerTools/Tests/Core/Classes/**/*'
        sp.framework = "XCTest"
        sp.requires_arc = true
        sp.dependency 'PowerTools/Core', '~> 0.4.1'
        sp.dependency 'PowerTools/CollectionVM', '~> 0.4.1'
    end
end
```

**Comment by vimaljake** [comment]
> So here's the thing.
> 
> I manually commented out the validation lines in `push.rb` just to get my private pod to push to our private spec repo. Everything works now and the pods work as expected.
> 
> I've since restored the `push.rb` back to its original state and tried pushing again to test whether the configuration options you suggested would fix the issue moving forward as I don't want to be editing Cocoapods source every time and now can't reproduce the error.
> 
> Not sure what's changed or what has fixed it but it's no longer complaining about the build not being compiled for testing.
> 
> Thanks for your help, and it's good to know that we can add build settings in the podspecs as this was something I was unsure about.

Can you show your podspec.

**Comment by sssandyad** [comment]
> Hi, I have the same issue, even adding the `spec.pod_target_xcconfig = { 'ENABLE_TESTABILITY' => 'YES' }`... This is my podspec, maybe someone could help me here:
> 
> ```ruby
> Pod::Spec.new do |s|
>     s.name             = 'PowerToolsTester'
>     s.version          = '0.4.1'
>     s.summary          = 'Tools tester for fast and modular app development.'
>     s.swift_version    = '5.0'
>     
>     s.description      = <<-DESC
>     Fast development tools with a better way to use UICollectionViews
>     DESC
>     
>     s.homepage         = 'https://github.com/Oni-zerone/PowerTools'
>     s.license          = { :type => 'MIT', :file => 'LICENSE' }
>     s.author           = { 'Andrea Altea' => 'oni.zerone@gmail.com' }
>     s.source           = { :git => 'https://github.com/Oni-zerone/PowerTools.git', :tag => s.version.to_s + '-tester' }
>     s.social_media_url = 'https://twitter.com/Oni_zerone'
>     s.pod_target_xcconfig = { 'ENABLED_TESTABILITY' => 'YES' }
>     
>     s.ios.deployment_target = '9.0'
>     
>     s.default_subspec = 'CoreTester'
>     s.subspec 'CoreTester' do |sp|
>         sp.source_files = 'PowerTools/Tests/Core/Classes/**/*'
>         sp.framework = "XCTest"
>         sp.requires_arc = true
>         sp.dependency 'PowerTools/Core', '~> 0.4.1'
>         sp.dependency 'PowerTools/CollectionVM', '~> 0.4.1'
>     end
> end
> ```

Use `ENABLE_TESTABILITY` instead of `ENABLED_TESTABILITY`

**Comment by nunogoncalves** [comment]
@cameroncooke hi! 
If I understand correctly you suggest adding the build setting ENABLE_TESTABILITY to the Sausage framework. Right? Doesn't that also set that to release builds?
